### PR TITLE
Use the get_attribute hook for dynamic classes

### LINKED ISSUE #6259: Call plugin get_attribute_hook() also for attributes that trigger __getattr__()
Currently, `get_attribute_hook()` is only called for attributes that are already present in the class symbol table. However, the typical use case is opposite -- dynamically added attributes. Therefore, the current behavior creates some complications while developing plugins, like every access to dynamic attribute first requires somehow placing it into the class symbol table using other hooks. I propose to widen the scope of `get_attribute_hook()` to at least be called for classes with `__getattr__()`.

### PR DESCRIPTION
**Pull request by lincolnq**
Whenever we find `__getattr__` or `__getattribute__` on an instance, the getattribute hook is called to give the hook a chance to modify the return type of one of those functions.

Closes #6259, #5910

(Note: Currently there are no tests. I can add them if important.)

### CODE REVIEW

File: mypy/checkmember.py [hunk 1]
```diff
@@ -340,7 +340,15 @@ def analyze_member_var_access(name: str,
                     typ = map_instance_to_supertype(itype, method.info)
                     getattr_type = expand_type_by_instance(bound_method, typ)
                     if isinstance(getattr_type, CallableType):
-                        return getattr_type.ret_type
+                        result = getattr_type.ret_type
+
+                        # call the get attribute hook before returning
```
Comment (ilevkivskyi) [inline mypy/checkmember.py]:
```suggestion
                        # Call the attribute hook before returning.
```

File: docs/source/extending_mypy.rst [hunk 2]
```diff
@@ -184,11 +184,7 @@ For example in this code:
 mypy will call ``get_method_signature_hook("ctypes.Array.__setitem__")``
 so that the plugin can mimic the ``ctypes`` auto-convert behavior.
 
-**get_attribute_hook** can be used to give more precise type of an instance
-attribute. Note however, that this method is only called for variables that
-already exist in the class symbol table. If you want to add some generated
-variables/methods to the symbol table you can use one of the three hooks
-below.
+**get_attribute_hook** overrides instance member field lookups and property access (not assignments, and not method calls). This hook is only called for fields which already exist on the class. *Exception:* if ``__getattr__`` or ``__getattribute__`` is a method on the class, the hook is called for all fields which do not refer to methods.
```
Comment (ilevkivskyi) [inline docs/source/extending_mypy.rst]: Please split this huge line into separate lines (around 80 columns per line).

**Comment by lincolnq** [comment]
@ilevkivskyi Thanks for the review! I made the changes you requested.

One of the test cases is that the attribute hook can produce an error. I invented my own error, "Field does not exist", but ideally I would test that you can fallback to the normal mypy error handling -- but I don't think you can access a MemberContext to call `mx.msg.has_no_attr`. Is the way I did it ok?

**Comment by lincolnq** [comment]
Hm, the old docstring was wrong too (it specifies that the hook will be
called with the base class name where the field is defined, but it is
called with the derived class name). I assume the documented behavior is
the desirable one?

File: test-data/unit/check-custom-plugin.test [hunk 3]
```diff
@@ -143,19 +143,49 @@ tmp/mypy.ini:2: error: Return value of "plugin" must be a subclass of "mypy.plug
...
+
+ds: DerivedSignal[Callable[[int], None]] = DerivedSignal()
+ds('') # E: Argument 1 has incompatible type "str"; expected "int"
 [file m.py]
 from typing import TypeVar, Generic, Callable
 T = TypeVar('T', bound=Callable[..., None])
 class Signal(Generic[T]):
     __call__: Callable[..., None]  # This type is replaced by the plugin
+
+class DerivedSignal(Signal, Generic[T]): ...
```
Comment (ilevkivskyi) [inline test-data/unit/check-custom-plugin.test]: I think this should be `Signal[T]`, this will fix the test.

**Comment by ilevkivskyi** [comment]
> I assume the documented behavior is the desirable one?

Yes, I think giving the name of the base class where `__getattr__()` was defined would be consistent. So you probably need to update your code. (The fix for the old test I posted above.)

File: test-data/unit/check-custom-plugin.test (hunk 3 above)
Comment (ilevkivskyi) [reply test-data/unit/check-custom-plugin.test]: Hm, it looks like variables (unlike methods) are passed with non-mapped instance type, so your fix is right. Anyway, it i better to change this to `Signal[T]` (currently it is equivalent to `Signal[Any]`).

**Pull request merged**